src/Stratis.Bitcoin.Features.MemoryPool/BlocksDisconnectedSignaled.cs

```diff
@@ -0,0 +1,60 @@
+﻿using System;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Signals;
+
+namespace Stratis.Bitcoin.Features.MemoryPool
+{
+    /// <summary>
+    /// Mempool observer on disconnected block notifications.
+    /// </summary>
+    public class BlocksDisconnectedSignaled : SignalObserver<Block>
+    {
+        private readonly IMempoolValidator mempoolValidator;
+        private readonly MempoolSchedulerLock mempoolLock;
+        private readonly ILogger logger;
+
+        public BlocksDisconnectedSignaled(IMempoolValidator mempoolValidator,
+            MempoolSchedulerLock mempoolLock, ILoggerFactory loggerFactory)
+        {
+            this.mempoolValidator = mempoolValidator;
+            this.mempoolLock = mempoolLock;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        protected override void OnNextCore(Block block)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(block), block.GetHash());
+
+            this.AddBackToMempoolAsync(block).GetAwaiter().GetResult();
+
+            this.logger.LogTrace("(-)");
+        }
+
+        public async Task AddBackToMempoolAsync(Block block)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(block), block.GetHash());
+
+            var state = new MempoolValidationState(true);
+
+            await this.mempoolLock.WriteAsync(async () =>
+            {
+                foreach (Transaction transaction in block.Transactions)
```

Comment:
Are you sure, you can process the transactions in this order? I think it should be processed in reverse order

Reply:
No I think this is correct, trx at the bottom may be spending trx at the top and trx at the top should be spending trx in the chain.

Would it make sense to look at the possibility to push back to mempool without full trx validation

Reply:
It turned out the question is trickier than it seems, I need to think about it little more

Reply:
As for the transaction validation, I'm thinking there could be problems

Reply:
I think you're right, it is because we do validation one by one and not as a batch, we need to go in this order and not in reverse order.